base/applications/rapps/geninst.cpp

```diff
@@ -0,0 +1,813 @@
+        narrow.Replace('/', '\\');
+        while (narrow[0] == '\\')
+            narrow = narrow.Mid(1);
+        UINT codepage = (fi.flag & pkzefsutf8) ? CP_UTF8 : 1252;
+        UINT cch = MultiByteToWideChar(codepage, 0, narrow, -1, NULL, 0);
+        cch = MultiByteToWideChar(codepage, 0, narrow, -1, path.GetBuffer(cch - 1), cch);
+        if (!cch)
+            break;
+        path.ReleaseBuffer(cch - 1);
+        DWORD fileatt = FILE_ATTRIBUTE_NORMAL, err;
+        BYTE attsys = HIBYTE(fi.version), dos = 0, ntfs = 10, vfat = 14;
+        if ((attsys == dos || attsys == ntfs || attsys == vfat) && LOBYTE(fi.external_fa))
+            fileatt = LOBYTE(fi.external_fa);
+
+        if (!NotifyFileExtractCallback(path, fi.uncompressed_size, fileatt,
+                                       Callback, Cookie))
+            continue; // Skip file
+
+        path = BuildPath(OutputDir, path);
+        SplitFileAndDirectory(path, &dir);
+        if (!dir.IsEmpty() && (err = CreateDirectoryTree(dir)))
+        {
+            zerr = UNZ_ERRNO;
+            SetLastError(err);
+            break;
+        }
+
+        if ((zerr = unzOpenCurrentFile(hzf)) != UNZ_OK)
+            break;
+        zerr = UNZ_ERRNO;
+        HANDLE hFile = CreateFileW(path, GENERIC_READ | GENERIC_WRITE,
+                                   FILE_SHARE_READ | FILE_SHARE_DELETE,
+                                   NULL, CREATE_ALWAYS, fileatt, NULL);
+        if (hFile != INVALID_HANDLE_VALUE)
+        {
+            DWORD len = 1024 * 4, cb;
+            LPSTR buf = narrow.GetBuffer(len);
+            for (zerr = UNZ_OK; zerr == UNZ_OK;)
+            {
+                len = zerr = unzReadCurrentFile(hzf, buf, len);
+                if (zerr <= 0)
+                    break;
+                zerr = WriteFile(hFile, buf, len, &cb, NULL) && cb == len ? UNZ_OK : UNZ_ERRNO;
+            }
+            CloseHandle(hFile);
+        }
+        unzCloseCurrentFile(hzf);
+    }
+    unzClose(hzf);
+    return zerr == UNZ_END_OF_LIST_OF_FILE ? UNZ_OK : zerr;
+}
+
+static UINT
+ExtractZip(LPCWSTR Archive, const CStringW &OutputDir,
+           EXTRACTCALLBACK Callback, void *Cookie)
+{
+    int zerr = ExtractFilesFromZip(Archive, OutputDir, Callback, Cookie);
+    return zerr == UNZ_ERRNO ? GetLastError() : zerr ? ERROR_INTERNAL_ERROR : 0;
+}
+
+static UINT
+ExtractCab(LPCWSTR Archive, const CStringW &OutputDir,
+           EXTRACTCALLBACK Callback, void *Cookie)
+{
+    if (ExtractFilesFromCab(Archive, OutputDir, Callback, Cookie))
+        return ERROR_SUCCESS;
+    UINT err = GetLastError();
+    return err ? err : ERROR_INTERNAL_ERROR;
+}
+
+enum { IM_STARTPROGRESS = WM_APP, IM_PROGRESS, IM_END };
+
+static struct CommonInfo
+{
+    LPCWSTR AppName;
+    HWND hDlg;
+    DWORD Error, Count;
+    BOOL Silent;
+    CRegKey *ArpKey, Entries;
+
+    CommonInfo(LPCWSTR DisplayName, BOOL IsSilent = FALSE)
+    : AppName(DisplayName), hDlg(NULL), Error(0), Count(0), Silent(IsSilent)
```

Comment:
You are not initializing `ArpKey` here.

Reply:
It's always initialized properly in the required places but OK.